src/main/java/com/google/sps/servlets/GmailActionableEmailsServlet.java

```diff
@@ -0,0 +1,133 @@
+ * email's subject line. Used by "Assign" panel on client.
+ */
+@WebServlet("/gmail-actionable-emails")
+public class GmailActionableEmailsServlet extends AuthenticatedHttpServlet {
+  private GmailClientFactory gmailClientFactory;
+
+  /** Create new servlet instance (called by java during HttpRequest handling */
+  public GmailActionableEmailsServlet() {
+    gmailClientFactory = new GmailClientImpl.Factory();
+  }
+
+  /**
+   * Create new servlet instance (used for testing)
+   *
+   * @param authenticationVerifier implementation of AuthenticationVerifier interface
+   * @param gmailClientFactory implementation of GmailClientFactory interface
+   */
+  public GmailActionableEmailsServlet(
+      AuthenticationVerifier authenticationVerifier, GmailClientFactory gmailClientFactory) {
+    super(authenticationVerifier);
+    this.gmailClientFactory = gmailClientFactory;
+  }
+
+  /**
+   * Retrieve the actionable emails from a user's gmail account, given a list of words/phrases to
+   * search for in the subject line.
+   *
+   * @param request HTTP request from client
+   * @param response Http response to be sent to client
+   * @param googleCredential valid, verified google credential object
+   * @throws IOException If a read/write issue arises while processing the request
+   */
+  @Override
+  public void doGet(
+      HttpServletRequest request, HttpServletResponse response, Credential googleCredential)
+      throws IOException {
+    GmailClient gmailClient = gmailClientFactory.getGmailClient(googleCredential);
+
+    List<String> subjectLinePhrases;
+    try {
+      subjectLinePhrases = getListFromQueryString(request, "subjectLinePhrases");
+    } catch (IllegalArgumentException e) {
+      response.sendError(400, "subjectLinePhrases must be present in request");
+      return;
+    }
+    if (subjectLinePhrases.isEmpty()) {
+      response.sendError(400, "subjectLinePhrases must be non-empty");
+      return;
+    }
+
+    boolean unreadOnly;
+    String unreadOnlyParameter = request.getParameter("unreadOnly");
+    if (unreadOnlyParameter == null) {
+      response.sendError(400, "unreadOnly must be present in request");
+      return;
+    }
+    if (!unreadOnlyParameter.equals("false") && !unreadOnlyParameter.equals("true")) {
+      response.sendError(400, "unreadOnly must be a boolean");
+    }
+    unreadOnly = Boolean.parseBoolean(unreadOnlyParameter);
+
+    int nDays;
+    try {
+      nDays = Integer.parseInt(request.getParameter("nDays"));
+    } catch (NumberFormatException e) {
+      response.sendError(400, "nDays must be present in request");
+      return;
+    }
+    if (nDays <= 0) {
+      response.sendError(400, "nDays must be positive");
+      return;
+    }
+
+    List<Message> actionableEmails =
+        gmailClient.getActionableEmails(
+            GmailClient.MessageFormat.FULL, subjectLinePhrases, unreadOnly, nDays);
+    JsonUtility.sendJson(response, actionableEmails);
```

Comment:
I think it's more efficient to just return the data that the client needs - even at the expense of some CPU work in the servlet. Also, querying METADATA without the body is probably more efficient on the Gmail side, and would be a win overall.

Reply:
Agreed- I made some changes to the Gmail client to facilitate this, as well as created a new POJO to ensure that only the correct data is being returned. 

Please let me know what you think - this was a substantial change.